File: src/main/java/net/ndrei/teslacorelib/gui/BasicContainerGuiPiece.java
package net.ndrei.teslacorelib.gui;

/**
 * Created by CF on 2016-12-21.
 */
public abstract class BasicContainerGuiPiece implements IGuiContainerPiece {
    private int left, top, width, height;
    private boolean isVisible = true;

    protected BasicContainerGuiPiece(int left, int top, int width, int height) {
        this.left = left;
        this.top = top;
        this.width = width;
        this.height = height;
    }

    @Override
    public int getLeft() {
        return this.left;
    }

    @Override
    public int getTop() {
        return this.top;
    }

    @Override
    public int getWidth() {
        return this.width;
    }

    @Override
    public int getHeight() {
        return this.height;
    }

    @Override
    public boolean isVisible() {
        return this.isVisible;
    }

    @Override
    public void setVisibility(boolean isVisible) {
        this.isVisible = isVisible;
    }

    @Override
    public void mouseClicked(BasicTeslaGuiContainer container, int mouseX, int mouseY, int mouseButton) {
    }

//    @Override
//    public void mouseClickMove(BasicTeslaGuiContainer container, int mouseX, int mouseY, int clickedMouseButton, long timeSinceLastClick) {
//    }

//    @Override
//    public void mouseReleased(BasicTeslaGuiContainer container, int mouseX, int mouseY, int state) {
//    }

    @Override
    public void drawBackgroundLayer(BasicTeslaGuiContainer container, int guiX, int guiY, float partialTicks, int mouseX, int mouseY) {
    }

    @Override
    public void drawMiddleLayer(BasicTeslaGuiContainer container, int guiX, int guiY, float partialTicks, int mouseX, int mouseY) {
    }

    @Override
    public void drawForegroundLayer(BasicTeslaGuiContainer container, int guiX, int guiY, int mouseX, int mouseY) {
    }

    @Override
    public void drawForegroundTopLayer(BasicTeslaGuiContainer container, int guiX, int guiY, int mouseX, int mouseY) {
    }

    public static boolean isInside(BasicTeslaGuiContainer container, IGuiContainerPiece piece, int mouseX, int mouseY) {
        int l = container.getGuiLeft() + piece.getLeft();
        if ((mouseX < l) || (mouseX > (l + piece.getWidth()))) {
            return false;
        }

        int t = container.getGuiTop() + piece.getTop();
        return ((mouseY >= t) && (mouseY <= (t + piece.getHeight())));
    }

    protected boolean isInside(BasicTeslaGuiContainer container, int mouseX, int mouseY) {
//        int l = container.getGuiLeft() + this.getLeft();
//        if ((mouseX < l) || (mouseX > (l + this.getWidth()))) {
//            return false;
//        }
//
//        int t = container.getGuiTop() + this.getTop();
//        return ((mouseY >= t) && (mouseY <= (t + this.getHeight())));
        return BasicContainerGuiPiece.isInside(container, this, mouseX, mouseY);
    }
}

File: src/main/kotlin/net/ndrei/teslacorelib/gui/BasicContainerGuiPiece.kt
package net.ndrei.teslacorelib.gui

/**
 * Created by CF on 2017-06-28.
 */
abstract class BasicContainerGuiPiece protected constructor(override val left: Int, top: Int, override val width: Int, override val height: Int) : IGuiContainerPiece {
    private var visible = true
    override var top: Int = top
        protected set

    override val isVisible get() = this.visible

    override fun setVisibility(isVisible: Boolean) {
        this.visible = isVisible
    }

    override fun mouseClicked(container: BasicTeslaGuiContainer<*>, mouseX: Int, mouseY: Int, mouseButton: Int) {}

    override fun drawBackgroundLayer(container: BasicTeslaGuiContainer<*>, guiX: Int, guiY: Int, partialTicks: Float, mouseX: Int, mouseY: Int) {}

    override fun drawMiddleLayer(container: BasicTeslaGuiContainer<*>, guiX: Int, guiY: Int, partialTicks: Float, mouseX: Int, mouseY: Int) {}

    override fun drawForegroundLayer(container: BasicTeslaGuiContainer<*>, guiX: Int, guiY: Int, mouseX: Int, mouseY: Int) {}

    override fun drawForegroundTopLayer(container: BasicTeslaGuiContainer<*>, guiX: Int, guiY: Int, mouseX: Int, mouseY: Int) {}

    protected fun isInside(container: BasicTeslaGuiContainer<*>, mouseX: Int, mouseY: Int): Boolean {
        return BasicContainerGuiPiece.isInside(container, this, mouseX, mouseY)
    }

    companion object {
        fun isInside(container: BasicTeslaGuiContainer<*>, piece: IGuiContainerPiece, mouseX: Int, mouseY: Int): Boolean {
            val l = container.guiLeft + piece.left
            if (mouseX < l || mouseX > l + piece.width) {
                return false
            }

            val t = container.guiTop + piece.top
            return mouseY >= t && mouseY <= t + piece.height
        }
    }
}
